src/test/basics/PerfLog_test.cpp

```diff
@@ -0,0 +1,1075 @@
+        std::shuffle (ids.begin(), ids.end(), shuffler_);
+
+        // Start all of the RPC commands twice to show they can all be tracked
+        // simultaneously.
+        for (int labelIndex = 0; labelIndex < labels.size(); ++labelIndex)
+        {
+            for (int idIndex = 0; idIndex < 2; ++idIndex)
+            {
+                std::this_thread::sleep_for (std::chrono::microseconds (10));
+                perfLog->rpcStart (
+                    labels[labelIndex], ids[(labelIndex * 2) + idIndex]);
+            }
+        }
+        {
+            // Examine current PerfLog::counterJson() values.
+            Json::Value const countersJson {perfLog->countersJson()["rpc"]};
+            BEAST_EXPECT(countersJson.size() == labels.size() + 1);
+            for (auto& label : labels)
+            {
+                // Expect every label in labels to have the same contents.
+                Json::Value const& counter {countersJson[label]};
+                BEAST_EXPECT(counter[jss::duration_us] == "0");
+                BEAST_EXPECT(counter[jss::errored] == "0");
+                BEAST_EXPECT(counter[jss::finished] == "0");
+                BEAST_EXPECT(counter[jss::started] == "2");
+            }
+            // Expect "total" to have a lot of "started"
+            Json::Value const& total {countersJson[jss::total]};
+            BEAST_EXPECT(total[jss::duration_us] == "0");
+            BEAST_EXPECT(total[jss::errored] == "0");
+            BEAST_EXPECT(total[jss::finished] == "0");
+            BEAST_EXPECT(jsonToUint64 (total[jss::started]) == ids.size());
+        }
+        {
+            // Verify that every entry in labels appears twice in currents.
+            // If we sort by duration_us they should be in the order the
+            // rpcStart() call was made.
+            std::vector<Cur> const currents {
+                getSortedCurrent (perfLog->currentJson()[jss::methods])};
+            BEAST_EXPECT(currents.size() == labels.size() * 2);
+
+            std::uint64_t prevDur = std::numeric_limits<std::uint64_t>::max();
+            for (int i = 0; i < currents.size(); ++i)
+            {
+                BEAST_EXPECT(currents[i].name == labels[i / 2]);
+                BEAST_EXPECT(prevDur > currents[i].dur);
+                prevDur = currents[i].dur;
+            }
+        }
+
+        // Finish all but the first RPC command in reverse order to show that
+        // the start and finish of the commands can interleave.  Half of the
+        // commands finish correctly, the other half with errors.
+        for (int labelIndex = labels.size() - 1; labelIndex > 0; --labelIndex)
+        {
+            std::this_thread::sleep_for (std::chrono::microseconds (10));
+            perfLog->rpcFinish (labels[labelIndex], ids[(labelIndex * 2) + 1]);
+            std::this_thread::sleep_for (std::chrono::microseconds (10));
+            perfLog->rpcError (labels[labelIndex], ids[(labelIndex * 2) + 0]);
```

Comment:
`+0` is a little odd. It doesn't really bother me, but it does seem a little odd.

Reply:
Inside this loop we use each label twice and each id once.  So we access the `ids` array with `(labelIndex * 2) + 1` and `(labelIndex * 2) + 0`.  The `+1` and the `+0` are intended to keep the constructs parallel in the programmers mind.